File: typescript/tests/core/hmac.test.ts
import { describe, expect, it } from 'bun:test';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';

import {
  AlgorithmRegistry,
  HMAC_MIN_SECRET_LEN,
  HMAC_SIG_LEN,
  hmacBackend,
} from '../../src/index.ts';

const FIXTURE_SECRET_PATH = join(import.meta.dir, '../../../fixtures/keys/hmac/secret.hex');

function readFixtureSecret(): Uint8Array {
  const hex = readFileSync(FIXTURE_SECRET_PATH, 'utf8').trim();
  return new Uint8Array(Buffer.from(hex, 'hex'));
}

describe('hmac backend — basic contract', () => {
  it('advertises alg = "hs256"', () => {
    expect(hmacBackend.alg).toBe('hs256');
  });

  it('registers into AlgorithmRegistry exactly once', () => {
    const reg = new AlgorithmRegistry();
    reg.register(hmacBackend);
    expect(() => reg.register(hmacBackend)).toThrow();
  });
});

describe('hmac backend — sign/verify', () => {
  it('round-trips with fixture secret', async () => {
    const secret = readFixtureSecret();
    expect(secret.length).toBe(HMAC_MIN_SECRET_LEN);
    const priv = await hmacBackend.importPrivate({ privateRaw: secret, publicRaw: secret });
    const pub = await hmacBackend.importPublic({ privateRaw: null, publicRaw: secret });
    const data = new TextEncoder().encode('hmac-round-trip');
    const sig = await hmacBackend.sign(priv, data);
    expect(sig).toBeInstanceOf(Uint8Array);
    expect(sig.length).toBe(HMAC_SIG_LEN);
    expect(await hmacBackend.verify(pub, data, sig)).toBe(true);
  });

  it('is deterministic (same input → same MAC)', async () => {
    const secret = readFixtureSecret();
    const k = await hmacBackend.importPrivate({ privateRaw: secret, publicRaw: secret });
    const data = new TextEncoder().encode('repeat-me');
    const a = await hmacBackend.sign(k, data);
    const b = await hmacBackend.sign(k, data);
    expect(a).toEqual(b);
  });

  it('rejects a tampered MAC', async () => {
    const secret = readFixtureSecret();
    const k = await hmacBackend.importPrivate({ privateRaw: secret, publicRaw: secret });
    const data = new TextEncoder().encode('tamper-test');
    const sig = await hmacBackend.sign(k, data);
    sig[0] = (sig[0] ?? 0) ^ 0xff;
    expect(await hmacBackend.verify(k, data, sig)).toBe(false);
  });

  it('rejects a wrong-length MAC without throwing', async () => {
    const secret = readFixtureSecret();
    const k = await hmacBackend.importPublic({ privateRaw: null, publicRaw: secret });
    expect(await hmacBackend.verify(k, new Uint8Array([1, 2]), new Uint8Array([1, 2, 3]))).toBe(
      false,
    );
  });

  it('different secrets do NOT verify each other', async () => {
    const secret1 = readFixtureSecret();
    const secret2 = new Uint8Array(32);
    for (let i = 0; i < 32; i++) secret2[i] = i;
    const k1 = await hmacBackend.importPrivate({ privateRaw: secret1, publicRaw: secret1 });
    const k2 = await hmacBackend.importPublic({ privateRaw: null, publicRaw: secret2 });
    const data = new TextEncoder().encode('cross-secret');
    const sig = await hmacBackend.sign(k1, data);
    expect(await hmacBackend.verify(k2, data, sig)).toBe(false);
  });
});

describe('hmac backend — minimum-strength enforcement', () => {
  it('rejects a 31-byte secret', async () => {
    await expect(
      hmacBackend.importPrivate({
        privateRaw: new Uint8Array(31),
        publicRaw: new Uint8Array(31),
      }),
    ).rejects.toThrow();
  });

  it('rejects PEM material (HMAC is raw-only)', async () => {
    await expect(
      hmacBackend.importPrivate({ privatePem: 'whatever', publicPem: 'whatever' }),
    ).rejects.toThrow();
    await expect(
      hmacBackend.importPublic({ privatePem: null, publicPem: 'whatever' }),
    ).rejects.toThrow();
  });
});

describe('hmac backend — generate()', () => {
  it('produces a ≥32-byte secret usable for sign/verify', async () => {
    const { raw } = await hmacBackend.generate('unused');
    expect(raw.privateRaw).not.toBeNull();
    expect(raw.privateRaw?.length).toBe(HMAC_MIN_SECRET_LEN);
    expect(raw.publicRaw.length).toBe(HMAC_MIN_SECRET_LEN);
    // Symmetric: privateRaw === publicRaw.
    expect(raw.privateRaw).toEqual(raw.publicRaw);
    const k = await hmacBackend.importPrivate(raw);
    const sig = await hmacBackend.sign(k, new Uint8Array([1, 2, 3]));
    expect(await hmacBackend.verify(k, new Uint8Array([1, 2, 3]), sig)).toBe(true);
  });
});

File: licensing/crypto/hmac/hmac_test.go
package hmac_test

import (
	"bytes"
	"encoding/hex"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"testing"

	lic "github.com/AnoRebel/licensing/licensing"
	hm "github.com/AnoRebel/licensing/licensing/crypto/hmac"
)

// repoRoot walks up until it finds fixtures/.
func repoRoot(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}
	for range 10 {
		if _, err := os.Stat(filepath.Join(dir, "fixtures", "keys")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	t.Fatalf("could not locate repo root from %s", dir)
	return ""
}

// -----------------------------------------------------------------------
// Round-trip
// -----------------------------------------------------------------------

func TestBackend_RoundTrip(t *testing.T) {
	be := hm.New()
	_, rawMat, err := be.Generate("")
	if err != nil {
		t.Fatalf("generate: %v", err)
	}
	if len(rawMat.PrivateRaw) != hm.DefaultSecretLen {
		t.Fatalf("generated secret len %d, want %d", len(rawMat.PrivateRaw), hm.DefaultSecretLen)
	}
	// PrivateRaw and PublicRaw should be value-equal for the symmetric case
	// but not alias each other.
	if !bytes.Equal(rawMat.PrivateRaw, rawMat.PublicRaw) {
		t.Fatal("generate: private and public raw secrets differ for HMAC")
	}
	rawMat.PrivateRaw[0] ^= 0xFF
	if bytes.Equal(rawMat.PrivateRaw, rawMat.PublicRaw) {
		t.Fatal("generate: PrivateRaw and PublicRaw alias the same backing array")
	}
	rawMat.PrivateRaw[0] ^= 0xFF // restore for the sign path below

	priv, err := be.ImportPrivate(lic.KeyMaterial{Raw: rawMat}, "")
	if err != nil {
		t.Fatal(err)
	}
	pub, err := be.ImportPublic(lic.KeyMaterial{Raw: rawMat})
	if err != nil {
		t.Fatal(err)
	}

	data := []byte("hello, licensing — hmac")
	sig, err := be.Sign(priv, data)
	if err != nil {
		t.Fatal(err)
	}
	if len(sig) != hm.SignatureLen {
		t.Fatalf("sig len %d, want %d", len(sig), hm.SignatureLen)
	}
	ok, err := be.Verify(pub, data, sig)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("verify reported false on a genuine HMAC")
	}

	// HMAC is deterministic — same secret + data → same MAC.
	sig2, _ := be.Sign(priv, data)
	if !bytes.Equal(sig, sig2) {
		t.Fatal("HMAC should be deterministic — two sigs over same data differed")
	}
}

// -----------------------------------------------------------------------
// Tampering
// -----------------------------------------------------------------------

func TestBackend_TamperedSignatureRejected(t *testing.T) {
	be := hm.New()
	_, rawMat, _ := be.Generate("")
	priv, _ := be.ImportPrivate(lic.KeyMaterial{Raw: rawMat}, "")
	pub, _ := be.ImportPublic(lic.KeyMaterial{Raw: rawMat})
	data := []byte("message")
	sig, _ := be.Sign(priv, data)

	tampered := bytes.Clone(sig)
	tampered[0] ^= 1
	ok, err := be.Verify(pub, data, tampered)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("tampered HMAC accepted")
	}

	// Wrong-length sig: (false, nil), not an error.
	short := sig[:16]
	ok, err = be.Verify(pub, data, short)
	if err != nil || ok {
		t.Fatalf("expected (false, nil) for short sig, got (%v, %v)", ok, err)
	}

	// Wrong data, same sig.
	ok, err = be.Verify(pub, []byte("different message"), sig)
	if err != nil || ok {
		t.Fatalf("verify with wrong data: ok=%v err=%v", ok, err)
	}
}

// -----------------------------------------------------------------------
// Key strength enforcement
// -----------------------------------------------------------------------

func TestImportPrivate_RejectsShortSecret(t *testing.T) {
	short := make([]byte, 16) // 16 bytes, below the 32-byte floor
	_, err := hm.New().ImportPrivate(lic.KeyMaterial{
		Raw: lic.RawKeyMaterial{PrivateRaw: short},
	}, "")
	if !errors.Is(err, lic.ErrInsufficientKeyStrength) {
		t.Fatalf("expected InsufficientKeyStrength, got %v", err)
	}
}

func TestImportPublic_RejectsShortSecret(t *testing.T) {
	short := make([]byte, 31) // one byte under the floor
	_, err := hm.New().ImportPublic(lic.KeyMaterial{
		Raw: lic.RawKeyMaterial{PublicRaw: short},
	})
	if !errors.Is(err, lic.ErrInsufficientKeyStrength) {
		t.Fatalf("expected InsufficientKeyStrength, got %v", err)
	}
}

func TestImport_RejectsPEMInput(t *testing.T) {
	_, err := hm.New().ImportPrivate(lic.KeyMaterial{
		Pem: lic.PemKeyMaterial{PrivatePem: "-----BEGIN PRIVATE KEY-----\nAA==\n-----END PRIVATE KEY-----"},
	}, "")
	if !errors.Is(err, lic.ErrTokenMalformed) {
		t.Fatalf("expected TokenMalformed for PEM-to-HMAC, got %v", err)
	}
	_, err = hm.New().ImportPublic(lic.KeyMaterial{
		Pem: lic.PemKeyMaterial{PublicPem: "-----BEGIN PUBLIC KEY-----\nAA==\n-----END PUBLIC KEY-----"},
	})
	if !errors.Is(err, lic.ErrTokenMalformed) {
		t.Fatalf("expected TokenMalformed for PEM-to-HMAC, got %v", err)
	}
}

func TestImport_MissingMaterial(t *testing.T) {
	_, err := hm.New().ImportPrivate(lic.KeyMaterial{}, "")
	if !errors.Is(err, lic.ErrTokenMalformed) {
		t.Fatalf("expected TokenMalformed, got %v", err)
	}
}

// -----------------------------------------------------------------------
// Secret accessor + isolation
// -----------------------------------------------------------------------

func TestSecret_Isolation(t *testing.T) {
	be := hm.New()
	_, rawMat, _ := be.Generate("")
	priv, _ := be.ImportPrivate(lic.KeyMaterial{Raw: rawMat}, "")
	got, err := hm.Secret(priv)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(got, rawMat.PublicRaw) {
		t.Fatal("Secret drift")
	}
	// Mutating the returned secret must not mutate the handle.
	got[0] ^= 0xFF
	got2, _ := hm.Secret(priv)
	if got[0] == got2[0] {
		t.Fatal("Secret returned a shared slice — mutation leaked")
	}
}

// -----------------------------------------------------------------------
// Fixture-backed interop
// -----------------------------------------------------------------------

func readHmacFixtureSecret(t *testing.T) []byte {
	t.Helper()
	root := repoRoot(t)
	b, err := os.ReadFile(filepath.Join(root, "fixtures", "keys", "hmac", "secret.hex"))
	if err != nil {
		t.Fatal(err)
	}
	raw, err := hex.DecodeString(strings.TrimSpace(string(b)))
	if err != nil {
		t.Fatal(err)
	}
	return raw
}

func TestBackend_ImportsFixtureSecret(t *testing.T) {
	be := hm.New()
	secret := readHmacFixtureSecret(t)
	priv, err := be.ImportPrivate(lic.KeyMaterial{
		Raw: lic.RawKeyMaterial{PrivateRaw: secret},
	}, "")
	if err != nil {
		t.Fatalf("import fixture secret (priv): %v", err)
	}
	pub, err := be.ImportPublic(lic.KeyMaterial{
		Raw: lic.RawKeyMaterial{PublicRaw: secret},
	})
	if err != nil {
		t.Fatalf("import fixture secret (pub): %v", err)
	}
	sig, err := be.Sign(priv, []byte("ping"))
	if err != nil {
		t.Fatal(err)
	}
	ok, err := be.Verify(pub, []byte("ping"), sig)
	if err != nil || !ok {
		t.Fatalf("fixture secret verify failed: ok=%v err=%v", ok, err)
	}
}

// Cross-language interop — TS-signed HMAC token verifies in Go.
func TestVerifyFixtureToken_003_HS256(t *testing.T) {
	root := repoRoot(t)
	name := "003-hs256-active"
	tokBytes, err := os.ReadFile(filepath.Join(root, "fixtures", "tokens", name, "expected_token.txt"))
	if err != nil {
		t.Fatal(err)
	}
	token := strings.TrimRight(string(tokBytes), "\n")

	secret := readHmacFixtureSecret(t)

	be := hm.New()
	reg := lic.NewAlgorithmRegistry()
	if err := reg.Register(be); err != nil {
		t.Fatal(err)
	}
	bindings := lic.NewKeyAlgBindings()
	if err := bindings.Bind("fixture-hmac-1", lic.AlgHS256); err != nil {
		t.Fatal(err)
	}
	keys := map[string]lic.KeyRecord{
		"fixture-hmac-1": {
			Kid: "fixture-hmac-1",
			Alg: lic.AlgHS256,
			Raw: lic.RawKeyMaterial{PublicRaw: secret},
		},
	}

	parts, err := lic.Verify(token, lic.VerifyOptions{
		Registry: reg,
		Bindings: bindings,
		Keys:     keys,
	})
	if err != nil {
		t.Fatalf("cross-language verify of TS-signed HMAC token FAILED: %v", err)
	}
	if parts.Header.Kid != "fixture-hmac-1" {
		t.Fatalf("header drift: %+v", parts.Header)
	}
	if parts.Payload["jti"] != "jti-active-0" {
		t.Fatalf("payload.jti drift: %v", parts.Payload["jti"])
	}
}
